test/core/iomgr/threadpool_test.cc

```diff
@@ -0,0 +1,187 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "src/core/lib/iomgr/executor/threadpool.h"
+
+#include "test/core/util/test_config.h"
+
+const int kSmallThreadPoolSize = 20;
+const int kLargeThreadPoolSize = 100;
+const int kThreadSmallIter = 100;
+const int kThreadLargeIter = 10000;
+
+// Simple functor for testing. It will count how many times being called.
+class SimpleFunctorForAdd : public grpc_experimental_completion_queue_functor {
+ public:
+  friend class SimpleFunctorCheckForAdd;
+  SimpleFunctorForAdd() {
+    functor_run = &SimpleFunctorForAdd::Run;
+    internal_next = this;
+    internal_success = 0;
+  }
+  ~SimpleFunctorForAdd() {}
+  static void Run(struct grpc_experimental_completion_queue_functor* cb,
+                  int ok) {
+    auto* callback = static_cast<SimpleFunctorForAdd*>(cb);
+    callback->count_.FetchAdd(1, grpc_core::MemoryOrder::RELAXED);
+  }
+
+  int count() { return count_.Load(grpc_core::MemoryOrder::RELAXED); }
+
+ private:
+  grpc_core::Atomic<int> count_{0};
+};
+
+// Checks the given SimpleFunctorForAdd's count with a given number.
+class SimpleFunctorCheckForAdd
+    : public grpc_experimental_completion_queue_functor {
+ public:
+  SimpleFunctorCheckForAdd(int ok, int* count) : count_(count) {
+    functor_run = &SimpleFunctorCheckForAdd::Run;
+    internal_success = ok;
+  }
+  ~SimpleFunctorCheckForAdd() {}
+  static void Run(struct grpc_experimental_completion_queue_functor* cb,
+                  int ok) {
+    auto* callback = static_cast<SimpleFunctorCheckForAdd*>(cb);
+    (*callback->count_)++;
+    GPR_ASSERT(*callback->count_ == callback->internal_success);
+  }
+
+ private:
+  int* count_;
+};
+
+static void test_add(void) {
+  gpr_log(GPR_INFO, "test_add");
+  grpc_core::ThreadPool* pool =
+      grpc_core::New<grpc_core::ThreadPool>(kSmallThreadPoolSize, "test_add");
+
+  SimpleFunctorForAdd* functor = grpc_core::New<SimpleFunctorForAdd>();
+  for (int i = 0; i < kThreadSmallIter; ++i) {
+    pool->Add(functor);
+  }
+  grpc_core::Delete(pool);
+  GPR_ASSERT(functor->count() == kThreadSmallIter);
+  grpc_core::Delete(functor);
+  gpr_log(GPR_DEBUG, "Done.");
+}
+
+// Thread that adds closures to pool
+class WorkThread {
```

Comment:
If WorkThread is irrelevant to test_constructor, it should be adjacent to test_multi_add.

Reply:
The same refactor applies to SimpleFunctorCheckForAdd.

Reply:
Do we need to move SimpleFunctorCheckForAdd?